lnwallet/musig_session.go

```diff
@@ -0,0 +1,535 @@
+package lnwallet
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+
+	"github.com/btcsuite/btcd/btcec/v2"
+	"github.com/btcsuite/btcd/btcec/v2/schnorr"
+	"github.com/btcsuite/btcd/btcec/v2/schnorr/musig2"
+	"github.com/btcsuite/btcd/txscript"
+	"github.com/btcsuite/btcd/wire"
+	"github.com/davecgh/go-spew/spew"
+	"github.com/lightningnetwork/lnd/input"
+	"github.com/lightningnetwork/lnd/keychain"
+	"github.com/lightningnetwork/lnd/lnwire"
+	"github.com/lightningnetwork/lnd/shachain"
+)
+
+// MusigPartialSig is a wrapper around the base musig2.PartialSignature type
+// that also includes information about the set of nonces used, and also the
+// signer. This allows us to implement the input.Signature interface, as that
+// requires the ability to perform abstract verification based on a public key.
+type MusigPartialSig struct {
+	// sig is the actual musig2 partial signature.
+	sig *musig2.PartialSignature
+
+	// signerNonce is the nonce used by the signer to generate the partial
+	// signature.
+	signerNonce [musig2.PubNonceSize]byte
+
+	// combinedNonce is the combined nonce of all signers.
+	combinedNonce [musig2.PubNonceSize]byte
+
+	// signerKeys is the set of public keys of all signers.
+	signerKeys []*btcec.PublicKey
+}
+
+// NewMusigPartialSig creates a new musig partial signature.
+func NewMusigPartialSig(sig *musig2.PartialSignature,
+	signerNonce, combinedNonce [musig2.PubNonceSize]byte,
+	signerKeys []*btcec.PublicKey) *MusigPartialSig {
+
+	return &MusigPartialSig{
+		sig:           sig,
+		signerNonce:   signerNonce,
+		combinedNonce: combinedNonce,
+		signerKeys:    signerKeys,
+	}
+}
+
+// FromWireSig maps a wire partial sig to this internal type that we'll use to
+// perform signature validation.
+func (p *MusigPartialSig) FromWireSig(sig *lnwire.PartialSigWithNonce,
+) *MusigPartialSig {
+
+	p.sig = &musig2.PartialSignature{
+		S: &sig.Sig,
+	}
+	p.signerNonce = sig.Nonce
+
+	return p
+}
+
+// ToWireSig maps the partial signature to something that we can use to write
+// out for the wire protocol.
+func (p *MusigPartialSig) ToWireSig() *lnwire.PartialSigWithNonce {
+	return &lnwire.PartialSigWithNonce{
+		PartialSig: lnwire.NewPartialSig(*p.sig.S),
+		Nonce:      p.signerNonce,
+	}
+}
+
+// Serialize serializes the musig2 partial signature. The serializing includes
+// the signer's public nonce _and_ the partial signature. The final signature
+// is always 98 bytes in length.
+func (p *MusigPartialSig) Serialize() []byte {
+	var b bytes.Buffer
+
+	p.ToWireSig().Encode(&b)
```

Comment:
can we be sure that this will never error? (I know that we cant return an error cause `Serialize` is part of the input.Signature)

I guess the layer above can also just check the number of bytes returned & assert that it is 98 vs 0 (which would mean error)